pkg/handler/github/github_hook.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright 2019 The Kubesphere Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package github
+
+import (
+	"context"
+	"fmt"
+	"github.com/emicklei/go-restful"
+	"github.com/google/go-github/github"
+	devopsv1alpha1 "github.com/kubesphere/s2ioperator/pkg/apis/devops/v1alpha1"
+	"io/ioutil"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	log "k8s.io/klog"
+	"net/http"
+	"regexp"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+)
+
+const (
+	s2irunCreatorPre = "trigger-"
+	pushEvent        = "push"
+)
+
+type Trigger struct {
+	KubeClientSet  client.Client
+	S2iBuilderName string
+	Namespace      string
+}
+
+func NewTrigger(client client.Client) *Trigger {
+	return &Trigger{
+		KubeClientSet: client,
+	}
+}
+
+func (g *Trigger) Serve(request *restful.Request, response *restful.Response) {
+	g.S2iBuilderName = request.PathParameter("s2ibuilder")
+	g.Namespace = request.PathParameter("namespace")
+
+	eventType := github.WebHookType(request.Request)
+	// Currently only accepting json payloads.
+	eventPayload, err := ioutil.ReadAll(request.Request.Body)
+	if err != nil {
+		log.Errorf("Error reading event body: %s", err)
+		response.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// validate payload
+	payload, err := g.ValidateTrigger(eventType, eventPayload)
+	if err != nil {
+		log.Error("Failed to validate event")
+		response.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	err = g.Action(eventType, payload)
+	if err != nil {
+		log.Error(err, "Failed to handle event")
+		response.WriteHeader(http.StatusInternalServerError)
+	}
+	response.WriteHeader(http.StatusCreated)
+	log.Infof("Github handing event with S2IBuilder name %s in namespace %s", g.S2iBuilderName, g.Namespace)
+
+}
+
+func (g *Trigger) ValidateTrigger(eventType string, payload []byte) ([]byte, error) {
+	instance := &devopsv1alpha1.S2iBuilder{}
+	namespacedName := &types.NamespacedName{Namespace: g.Namespace, Name: g.S2iBuilderName}
+	err := g.KubeClientSet.Get(context.TODO(), *namespacedName, instance)
+	if err != nil {
+		log.Errorf("Failed to get S2IBuilder: %s, in namespace %s, with error: %s", g.S2iBuilderName, g.Namespace, err)
+		return nil, err
+	}
+
+	// Check if the event type is in the allow-list, Now just support push event.
+	if eventType != pushEvent {
+		return nil, fmt.Errorf("not support event type %s", eventType)
+	}
+	if instance.Spec.Config.AllowedEvents != nil {
```

Comment:
We only parse push events, do we still need this field?

Reply:
this field is for extend in future. it also can delete.